### PR TITLE
Add more failure policy integration tests and refactor into table driven format

### PR DESCRIPTION
**Pull request by danielvegamyhre**
- Adds more failure policy integration tests and refactor into table driven format.
- Fixes a bug in the jobset controller causing it to try to recreate existing jobs marked for deletion that had not yet been deleted
- Refactor `constructJobsFromTemplate` method of `JobSetReconciler` to make it more concise and readable.

### CODE REVIEW

File: test/integration/jobset_controller_test.go [hunk 1]
```diff
@@ -371,48 +382,115 @@ var _ = ginkgo.Describe("JobSet controller", func() {
...
+			},
+			{
+				jobUpdateFn: func(jobList *batchv1.JobList) error {
+					ginkgo.By("failing another job")
+					job := &jobList.Items[0]
+					job.Status.Conditions = append(job.Status.Conditions, batchv1.JobCondition{
+						Type:   batchv1.JobFailed,
+						Status: corev1.ConditionTrue,
+					})
+					gomega.Expect(k8sClient.Status().Update(ctx, job)).Should(gomega.Succeed())
```
Comment (ahg-g) [inline test/integration/jobset_controller_test.go:444]: put this in a "failJob" function

File: test/integration/jobset_controller_test.go [hunk 2]
```diff
@@ -371,48 +382,115 @@ var _ = ginkgo.Describe("JobSet controller", func() {
...
+		ginkgo.Entry("1 job succeeds 1 job fails, all jobs recreated", []*jobUpdate{
+			{
+				jobUpdateFn: func(jobList *batchv1.JobList) error {
+					ginkgo.By("succeeding a job")
+					job := jobList.Items[0]
+					job.Status.Conditions = append(job.Status.Conditions, batchv1.JobCondition{
+						Type:   batchv1.JobComplete,
+						Status: corev1.ConditionTrue,
+					})
+					gomega.Expect(k8sClient.Status().Update(ctx, &job)).Should(gomega.Succeed())
```
Comment (ahg-g) [inline test/integration/jobset_controller_test.go:459]: put this in a `completeJob` func

File: test/integration/jobset_controller_test.go [hunk 3]
```diff
@@ -426,3 +247,163 @@ func checkJobSetStatus(js *jobset.JobSet, condition jobset.JobSetConditionType)
...
+			return -1, err
+		}
+		return len(svcList.Items), nil
+	}).Should(gomega.Equal(numExpectedJobs(js)))
+}
+
+// - 3 replicated jobs, 1 replica each
+func simpleJobSet() *jobset.JobSet {
+	return testing.MakeJobSet("js-succeed", ns.Name).
+		AddReplicatedJob(testing.MakeReplicatedJob("replicated-job-a").
```
Comment (ahg-g) [inline test/integration/jobset_controller_test.go:328]: to simplify things and make sure we test cases where replicas > 1, lets have two replicatedJobs here, one with replicas = 1 and one with replicas > 1 and use that in all tests.

File: test/integration/jobset_controller_test.go [hunk 4]
```diff
@@ -426,3 +247,163 @@ func checkJobSetStatus(js *jobset.JobSet, condition jobset.JobSetConditionType)
...
+		var svcList corev1.ServiceList
+		if err := k8sClient.List(ctx, &svcList, client.InNamespace(js.Namespace)); err != nil {
+			return -1, err
+		}
+		return len(svcList.Items), nil
+	}).Should(gomega.Equal(numExpectedJobs(js)))
+}
+
+// - 3 replicated jobs, 1 replica each
+func simpleJobSet() *jobset.JobSet {
```
Comment (ahg-g) [inline test/integration/jobset_controller_test.go:326]: return a wrapper so we can use it to add FailurePolicy

File: test/integration/jobset_controller_test.go [hunk 5]
```diff
@@ -426,3 +247,163 @@ func checkJobSetStatus(js *jobset.JobSet, condition jobset.JobSetConditionType)
...
+			SetJob(testing.MakeJob("test-job", ns.Name).SetCompletionMode(batchv1.IndexedCompletion).Obj()).
+			SetEnableDNSHostnames(true).
+			Obj()).
+		Obj()
+}
+
+// - 2 replicated jobs, 3 replicas each
+// - Indexed completion mode
+// - Failure Policy of (operator='Any', restartPolicy='RecreateAll', maxRestarts=1)
+func jobSetFailurePolicyAnyRecreateAll() *jobset.JobSet {
```
Comment (ahg-g) [inline test/integration/jobset_controller_test.go:362]: we can remove this func, the test entry can call:

```
simpleJobSet().SetFailurePolicy(&jobset.FailurePolicy{
			Operator:      jobset.TerminationPolicyTargetAny,
			RestartPolicy: jobset.RestartPolicyRecreateAll,
			MaxRestarts:   1,
		}).
		Obj()
```

**Comment by ahg-g** [comment]
> Fixes a bug in the jobset controller causing it to try to recreate existing jobs marked for deletion that had not yet been deleted

Is this really a bug? perhaps we will get creation errors, but I would expect the jobs will eventually be created (the reconciler will get the delete event and that should allow recreating the job successfully)?

**Comment by danielvegamyhre** [comment]
> Is this really a bug? perhaps we will get creation errors, but I would expect the jobs will eventually be created (the reconciler will get the delete event and that should allow recreating the job successfully)?

It does seem like a bug to me even though it doesn't break anything and the jobs are eventually recreated correctly. The previous behavior was sending 2 job creation requests per deleted job: once when it is marked for deletion but still exists (this call will fail),  then once after the deletion occurs and another reconciliation is triggered and it sees the job does not exist and recreates it (this call succeeds). 

My thinking was since we know the first call will alway fail, i don't see a reason to spam the apiserver with bad requests and pollute the logs with error messages that may confuse/mislead users. What do you think?

**Comment by ahg-g** [comment]
A bug means the controller is not behaving as expected, which is not the case here (because the job is eventually created at the right time), but that doesn't mean we shouldn't optimize away those unnecessary calls. 

The reason I asked was mainly to double check my assumptions about the events the controller gets and how it reconciles them.

File: test/integration/jobset_controller_test.go [hunk 6]
```diff
@@ -36,383 +38,202 @@ const (
...
+				{
+					jobUpdateFn:             completeAllJobs,
+					expectedJobSetCondition: jobset.JobSetCompleted,
+				},
+			},
+		}),
+		ginkgo.Entry("jobset validation should fail if DNS hostnames is enabled and job completion mode is not indexed", &testCase{
+			makeJobSet:               jobSetWithDNSHostnamesEnabledNonIndexed,
+			jobSetCreationShouldFail: true,
+		}),
```
[the remark this answers is not in the document]
Comment (danielvegamyhre) [reply test/integration/jobset_controller_test.go]: Moved to top of table and added TODO to create separate Describe+DescribeTable for validation tests.

File: test/integration/jobset_controller_test.go (hunk 3 above)
Comment (danielvegamyhre) [reply test/integration/jobset_controller_test.go:356]: Ok the params for this base test jobset are now:

- one replicated job with 1 replica
- one replicated job with 3 replicas and pod DNS hostnames enabled

**Comment by danielvegamyhre** [comment]
@ahg-g I see, makes sense. Anyway I have finished addressing your comments and this is ready for another look

File: test/integration/jobset_controller_test.go [hunk 7]
```diff
@@ -113,8 +130,19 @@ var _ = ginkgo.Describe("JobSet controller", func() {
 				}
 			}
 		},
+		// TODO: move validation tests to separate Describe + DescribeTable
+		ginkgo.Entry("jobset validation should fail if DNS hostnames is enabled and job completion mode is not indexed", &testCase{
```
Comment (ahg-g) [inline test/integration/jobset_controller_test.go:134]: "validate enableDNSHostnames can't be set if job is not Indexed"

**Review by ahg-g — COMMENTED** [review]
one minor suggestion

File: test/integration/jobset_controller_test.go (hunk 7 above)
Comment (danielvegamyhre) [reply test/integration/jobset_controller_test.go]: Done

**Pull request merged**